src/hotspot/share/opto/memnode.hpp

```diff
@@ -1184,7 +1184,8 @@ class MemBarNode: public MultiNode {
     TrailingStore,
     LeadingStore,
     TrailingLoadStore,
-    LeadingLoadStore
+    LeadingLoadStore,
+    AfterPartialArrayCopy
```

Comment:
Idea here is to prevent bypassing arraycopy operation post expansion during memory chain discovery [and] optimization. Currently a memory barrier is inserted after array copy macro expansion into a stub call, this pattern is being checked during memory chain discovery,  with partial in-lining we create additional control structure for selection b/w slow path (stub call) and fast path (partially in-lined code).  To prevent increasing the complexity of patter matching introduced a flag in MemBarrier node which is set only if partial in-lining takes place.